### PR TITLE
feat: add bisection method

### PR DESCRIPTION
**Pull request by aybarsnazlica**
#### Description of Change

Added bisection method used for root finding. 

#### Checklist

- [x] Added description of change
- [x] Added file name matches [File name guidelines](https://github.com/TheAlgorithms/C/blob/master/CONTRIBUTING.md#File-Name-guidelines)
- [x] Added tests and example, test must pass
- [x] Relevant documentation/comments is changed or added
- [x] PR title follows semantic [commit guidelines](https://github.com/TheAlgorithms/C/blob/master/CONTRIBUTING.md#Commit-Guidelines)
- [x] Search previous suggestions before making a new one, as yours may be a duplicate.
- [x] I acknowledge that all my contributions will be made under the project's license.

Notes:

### CODE REVIEW

File: numerical_methods/bisection_method.c [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+{
+    int n = 1;     // step counter
+    float middle;  // midpoint
+
+    while (n <= NMAX)
+    {
+        middle = (x_left + x_right) / 2;  // bisect the interval
+        float error = fabs(x_left - x_right) / 2;
+
+        if (func(middle) == 0.0 || error < tolerance)
```
[the remark this answers is not in the document]
Comment (aybarsnazlica) [reply numerical_methods/bisection_method.c:59]: That's right. But I think that will hurt code readability.

File: numerical_methods/bisection_method.c [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+    {
+        middle = (x_left + x_right) / 2;  // bisect the interval
+        float error = fabs(x_left - x_right) / 2;
+
+        if (fabs(func(middle)) <= EPSILON || error < tolerance)
+        {
+            return middle;
+        }
+
+        if (sign(func(middle), func(x_left)) >= 0.0)
```
Comment (CascadingCascade) [inline numerical_methods/bisection_method.c:64]: You probably shouldn't use `>=` here, `func(x_left)` simply can't be zero, and if `func(middle)` equals zero the function would have returned already.

**Review by CascadingCascade — COMMENTED** [review]
I would suggest you to replace all `float` with `double`, more precision wouldn't hurt

File: numerical_methods/bisection_method.c [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+ */
+float bisection(float x_left, float x_right, float tolerance)
+{
+    int n = 1;     // step counter
+    float middle;  // midpoint
+
+    while (n <= NMAX)
+    {
+        middle = (x_left + x_right) / 2;  // bisect the interval
+        float error = fabs(x_left - x_right) / 2;
```
[the remark this answers is not in the document]
Comment (aybarsnazlica) [reply numerical_methods/bisection_method.c]: I see. Changed to `middle - x_left`.

File: numerical_methods/bisection_method.c [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+ * defined by these values and then selecting the subinterval in which the
+ * function changes sign, and therefore must contain a root. It is a very
+ * simple and robust method, but it is also relatively slow. Because of this,
+ * it is often used to obtain a rough approximation to a solution which is
+ * then used as a starting point for more rapidly converging methods.
+ * @author [Aybars Nazlica](https://github.com/aybarsnazlica)
+ */
+
+#include <assert.h>  /// for assert
+#include <math.h>    /// for pow and fabs
```
[the remark this answers is not in the document]
Comment (aybarsnazlica) [reply numerical_methods/bisection_method.c]: Right. I kept fabs for the tests.

**Review by alexpantyukhin — DISMISSED** [review]

**Review by CascadingCascade — DISMISSED** [review]
LGTM

File: numerical_methods/bisection_method.c [hunk 5]
```diff
@@ -0,0 +1,104 @@
+/**
+ * @file
+ * @brief In mathematics, the bisection method is a root-finding method that
+ * applies to any continuous function for which one knows two values with
+ * opposite signs. https://en.wikipedia.org/wiki/Bisection_method
```
Comment (Panquesito7) [inline numerical_methods/bisection_method.c:5]:
```suggestion
 * @brief In mathematics, the [Bisection Method](https://en.wikipedia.org/wiki/Bisection_method) is a root-finding method that
 * applies to any continuous function for which one knows two values with
 * opposite signs.
```

File: numerical_methods/bisection_method.c [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+ */
+double sign(double a, double b)
+{
+    return (a > 0 && b > 0) + (a < 0 && b < 0) - (a > 0 && b < 0) -
+           (a < 0 && b > 0);
+}
+
+/**
+ * @brief Polynomial function for root finding
+ * @param x
```
Comment (Panquesito7) [inline numerical_methods/bisection_method.c:36]: Undocumented parameter.

**Review by Panquesito7 — CHANGES_REQUESTED** [review]

File: numerical_methods/bisection_method.c (hunk 6 above)
Comment (aybarsnazlica) [reply numerical_methods/bisection_method.c]: Added documentation.

File: numerical_methods/bisection_method.c [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+#include <math.h>    /// for fabs
+
+#define EPSILON 0.0001  // a small positive infinitesimal quantity
+#define NMAX 50         // maximum number of iterations
+
+/**
+ * @brief Function to check if two values have the same sign
+ * @param a First value
+ * @param b Second value
+ * @returns `sign`
```
Comment (Panquesito7) [inline numerical_methods/bisection_method.c:27]: What is `sign` exactly? Please document this here.

File: numerical_methods/bisection_method.c [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+ * defined by the two values and then selecting the subinterval in which the
+ * function changes sign, and therefore must contain a root. It is a very
+ * simple and robust method, but it is also relatively slow. Because of this,
+ * it is often used to obtain a rough approximation to a solution which is
+ * then used as a starting point for more rapidly converging methods.
+ * @author [Aybars Nazlica](https://github.com/aybarsnazlica)
+ */
+
+#include <assert.h>  /// for assert
+#include <math.h>    /// for fabs
```
Comment (Panquesito7) [inline numerical_methods/bisection_method.c:18]: Needed for the `printf` function.
```suggestion
#include <math.h>    /// for fabs
#include <stdio.h>   /// for IO operations
```

File: numerical_methods/bisection_method.c [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+ * @returns void
+ */
+static void test()
+{
+    /* Compares root value that is found by the bisection method within a given
+     * floating point error*/
+    assert(fabs(bisection(1.0, 2.0, 0.0001) - 1.847473) <
+           EPSILON);  // the algorithm works as expected
+    assert(fabs(bisection(100.0, 250.0, 0.0001) - 249.999928) <
+           EPSILON);  // the algorithm works as expected
```
Comment (Panquesito7) [inline numerical_methods/bisection_method.c:95]:
```suggestion
           EPSILON);  // the algorithm works as expected

           printf("All tests have successfully passed!\n");
```

**Review by Panquesito7 — APPROVED** [review]
Thanks! 🚀

**Review by tjgurwara99 — APPROVED** [review]

**Review by alexpantyukhin — APPROVED** [review]

**Pull request merged**